Add Panoptes authentication

zwolf opened the pull request:
Adds parsing of Panoptes JWTs. Includes parsing requests for project roles and getting project information. ApplicationController bits for asking about users. User#from_jwt to pull the id/dname/admin info out. Specs for everything.

Requests should include the user's JWT token (`Bearer eysdf...`) in the `Authorization` header. This is parsed by the ApplicationController, where a current user is set. The user's roles are pulled from Panoptes `project_roles` and point and parsed. 

Next step will be to use these parsed project roles to implement Pundit policies.

zwolf commented:
Resolves https://github.com/zooniverse/tove/issues/7

camallen commented on lib/panoptes_client.rb:
```diff
@@ -0,0 +1,43 @@
+  def client
+    @client ||= Panoptes::Client.new({
+      env: env,
+      public_key_path: Authenticator.key_path,
```
The underlying client can handle this for you, https://github.com/zooniverse/panoptes-client.rb/blob/3bd483c20a94a70fac438de0fbe0661faa9b34cb/lib/panoptes/client.rb#L55

Most likely YAGNI but is there a user case where i'd like to test the tove staging app against the production API, i.e. configure these via a query param?

zwolf replied:
I like the idea of the public key handled by the gem. If this line can just be scrapped, I'll remove it.

zwolf replied:
This will happen with the Authenticator removal refactor.

camallen commented on app/controllers/application_controller.rb:
```diff
@@ -13,4 +18,23 @@ def jsonapi_render(collection, filters)
+
+  def auth_token
+    return @auth_token if @auth_token
+    authorization = request.headers['Authorization']
```
super minor feel free to dismiss, you can use the assignment expression to capture the clause on `auth.present?` and save an operation / branch test.
```
if (authorization = request.headers['Authorization'])
  @auth_token = authorization.sub(/^Bearer /, '')
end
```

I would probably just collapse them all to a regex with a capture group
`@auth_token = authorization.match(/\ABearer\s(.+)\z/)[1] # 2nd index is the match capture group`

camallen commented on spec/lib/panoptes_client_spec.rb:
```diff
@@ -0,0 +1,33 @@
+    let(:parsed_roles) { { 1 => ["collaborator"], 2 => ["owner"], 3 => ["researcher"] } }
+
+    it 'should return a hash' do
+      allow(client_double).to receive(:paginate).and_return(raw_roles.with_indifferent_access)
```
If the client only returns a hash with string keys (not indifferent) could this be introducing bugs when looking up response objects with symbol keys?

If it doesn't introduce bugs, could the `raw_roles` fixture ensure `with_indifferent_access` instead of everywhere you use this? Might save some extra typing and spec bugs where it's left off and we use a symbol vs string key

zwolf replied:
Can you control how Rspec parses the shared context? The fixtures have string keys, but after they're loaded by rspec they end up as symbols, which is why I had to add the indifferent_access bit.

camallen replied:
This is how ruby handles hashes now, you could add `with_indifferent_access` on the let block, e.g. here https://github.com/zooniverse/tove/blob/0809cc8488c6fef9544c9dc0be02f81abbd1a201/spec/fixtures/project_roles.rb#L22

or maybe use json to encode them across (multi line stings) but this is ugly, i'm not sure.

zwolf commented on app/models/user.rb:
```diff
@@ -1,4 +1,18 @@
+
+    User.where(id: id, login: login).first_or_create do |user|
+      user.display_name = data['dname']
+      user.admin = data['admin'] == true # explicit
```
Explicitly checking the JWT for the encoded admin flag.

camallen replied:
```suggestion
      # explicit set the internal admin flag if the user is an admin
      user.admin = data['admin'] == true 
```
What happens longer term, where say an admin leaves and loses admin status in main API? they'll still have admin rights on this system, right?

zwolf replied:
No, because this system would be parsing their new JWT forwarded by the front end and it wouldn't have the admin flag set. This is per request.

camallen replied:
Does `first_or_create` run the block on each request or only when the record is first created?

zwolf replied:
For the record (and in case this was a genuine question), the block isn't run if the record isn't created. Added `.tap` to the line to always run the block.

zwolf commented on lib/authenticator.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def self.from_token(token)
+    decode(token)&.dig 0, 'data'
```
Because it didn't have those features the last time I built something like this and it is entirely undocumented. I did end up in the source for some other reason and noticed that PR that was merged in March that added a few of these convenience methods, but this was almost done at that point.

camallen replied:
The library methods were added as work in caesar in march, https://github.com/zooniverse/caesar/issues/598

And this tove code was written in march? 

So.... the plan is to avoid using the tested / approved library code, instead you'll roll your own to handle this?

zwolf replied:
I'm going to open an issue regarding a refactor.

zwolf commented on lib/panoptes_client.rb:
```diff
@@ -0,0 +1,43 @@
+    client
+  end
+
+  def roles(user_id)
```
OK, I'll open an issue to extract it out.

camallen replied:
Do you have time to add the features to the `panoptes-client.rb`code base or will you file an issue and leave this in?

zwolf replied:
I'll open an issue now to extract it out later.

zwolf commented on app/models/user.rb:
```diff
@@ -1,4 +1,18 @@
+  end
+
+  def roles
+    @roles || { }
```
What if the user had their admin access revoked? Implementing storage+caching/expiration of tokens is something I can open an issue for and we can discuss there.

camallen replied:
sure if they have revoked roles caching is an issue. I think you'll see varying levels of performance when trying to enumerate these roles, especially for users with lots of roles and specifically when accessing these roles without the project context.

This can be dealt with outside the scope of this PR, please do file and issue for it and test for it.

zwolf replied:
https://github.com/zooniverse/tove/issues/27

zwolf commented on app/controllers/application_controller.rb:
```diff
@@ -13,4 +18,23 @@ def jsonapi_render(collection, filters)
+    @current_user = User.from_jwt Authenticator.from_token auth_token
+  end
+
+  def set_roles
```
It will be, when role-based permissions are implemented on the individual controllers.

zwolf commented on lib/panoptes_client.rb:
```diff
@@ -0,0 +1,43 @@
+
+  def initialize(token)
+    @token ||= token
+    client
```
You mean include a `client` instance in the initialize method signature here instead of building it via a PanoptesClient instance method? Happy to take a look at a specific suggestion and a case for why it's ideal, please open an issue.

zwolf commented on app/models/user.rb:
```diff
@@ -1,4 +1,18 @@
   validates :login, presence: true
+
+  def self.from_jwt(data = { })
+    id, login = data&.values_at 'id', 'login'
```
The token's being generated by Panoptes, so probably not. This method also handles the anonymous user case. I suppose it could if the JWT is edited manually by either the user or the front end viewer/editor, but it'd be a bit of work to edit that'd just end up with a valid request that has no permissions.

That said, it does seem like it should be the controller conditionally setting `@current_user` based on whether the token exists, and not the responsibility of `User#from_jwt` to decide. So I moved the token.present? question to `ApplicationController#set_user` and refactored that JWT method to throw an error if it gets called with bad data.

zwolf commented:
I've addressed most comments and created a number of issues related to others. I am merging this now and I will revisit individual concerns via those issues.
